system-tests/lib/cre/don/secrets/secrets.go

```diff
@@ -48,3 +56,243 @@ func GenerateSecrets(input *cretypes.GenerateSecretsInput) (cretypes.NodeIndexTo
+				return nil, fmt.Errorf("number of EVM keys for DON %d differs between chain IDs %d and %d", donMetadata.ID, firstChainID, chainID)
+			}
+			for i := range evmKeys.PublicAddresses {
+				if evmKeys.PublicAddresses[i] != firstEVMKeys.PublicAddresses[i] {
+					return nil, fmt.Errorf("EVM public address mismatch for DON %d, node %d between chain IDs %d and %d", donMetadata.ID, i, firstChainID, chainID)
+				}
+			}
+		}
+
+		// Now add the EVM addresses to the node metadata
+		for chainID, evmKeys := range chainIDsToEVMKeys {
+			if len(evmKeys.PublicAddresses) != len(donMetadata.NodesMetadata) {
+				return nil, fmt.Errorf("number of EVM keys for DON %d and chain ID %d does not match the number of nodes. Expected %d, got %d", donMetadata.ID, chainID, len(donMetadata.NodesMetadata), len(evmKeys.PublicAddresses))
+			}
+			for idx, nodeMetadata := range donMetadata.NodesMetadata {
+				nodeMetadata.Labels = append(nodeMetadata.Labels, &cretypes.Label{
+					Key:   node.EthAddressKey,
+					Value: evmKeys.PublicAddresses[idx].Hex(),
+				})
+			}
+			// Use first ETH address for the DON metadata, because all of them are the same
+			break
+		}
+	}
+
+	return topology, nil
+}
+
+// struct required for reading "address" from this bit of encrypted JSON:
+// JSON = '{"address":"e753ac0b6e175ce3a939c55433a0109c5a6f8777"}'
+type evmJSON struct {
+	Address string `json:"address"`
+}
+
+var publicEVMAddressFromEncryptedJSON = func(jsonString string) (string, error) {
+	var eJSON evmJSON
+	err := json.Unmarshal([]byte(jsonString), &eJSON)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to unmarshal evm json")
+	}
+	return eJSON.Address, nil
+}
+
+// struct required for reading "peerID" from this bit of encrypted JSON:
+// JSON = '{"keyType":"P2P","publicKey":"f3c458c9064bdde449a3904ba8d3f8f5ebf79623077430325252c3368f920199","peerID":"p2p_12D3KooWSDvtYVF3FoyGeMrmDxYeJZMzbEyMHRwmf5GUSqgJhST2"}'
+type p2pJSON struct {
+	PeerID string `json:"peerID"`
+}
+
+var publicP2PAddressFromEncryptedJSON = func(jsonString string) (string, error) {
+	var pJSON p2pJSON
+	err := json.Unmarshal([]byte(jsonString), &pJSON)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to unmarshal p2p json")
+	}
+	return pJSON.PeerID, nil
+}
+
+func KeysOutputFromConfig(nodeSets []*cretypes.CapabilitiesAwareNodeSet) (*cretypes.GenerateKeysOutput, error) {
+	output := &cretypes.GenerateKeysOutput{
+		EVMKeys: make(cretypes.DonsToEVMKeys),
+		P2PKeys: make(cretypes.DonsToP2PKeys),
+	}
+	p2pKeysFoundPerDon := make(map[uint32]int)
+	evmKeysFoundPerDon := make(map[uint32]int)
+	for donIdx, nodeSet := range nodeSets {
+		donIdxUint32 := uint32(donIdx) // #nosec G115: ignore as this will NEVER happen, we don't have zillions of DONs
+		p2pKeys := types.P2PKeys{}
+		evmKeysPerChainID := make(cretypes.ChainIDToEVMKeys)
+		for nodeIdx, nodeSpec := range nodeSet.NodeSpecs {
+			if nodeSpec.Node.TestSecretsOverrides != "" {
+				var secrets coreconfig.Secrets
```

Comment:
Can we duplicate this struct to break "chainlink/v2" dependency?

Reply:
done